### ISSUE TITLE
[Feature Request] An option to make text only match a certain rule

### ISSUE THREAD

**Issue by georgetzlp** [dlevs/vscode-regex-robin#3]
Currently using the following rules
```json
  "regexrobin.rules": [
    {
      "regex": "TODO:\\s+(\\w+)",
      "editor": [
        {
          "color": "orange",
          "hoverMessage": "$1"
        }
      ]
    },
    {
      "regex": "\\w+:\\s+(?=(\\w+))",
      "editor": [
        {
          "color": "cyan",
          "hoverMessage": "$1"
        }
      ]
    }
  ],
```
with the following text
```js
// TODO: test
```
would match both rules and the only way I could see to fix this would be to to change the second rule's regex to `\\w+(?<!TODO):\\s+(?=(\\w+))` which could get ugly fast.

As a solution I propose adding an option such as `override` or `ignoreOthers` could be useful. Another idea I had is the ability to add an importance value to each rule with the highest value being applied over all the other ones overwriting them.

This sounds quite tricky to implement; perhaps a temporary solution could be an option to allow a rule to take over an entire line, overwriting any other rule in that same line?

I understand if this isn't something you're willing to implement because this could very much be me being too picky with regex but I think it would be a nice addition.

**Comment by dlevs** [comment]
Hey, thanks for sharing!

> Another idea I had is the ability to add an importance value to each rule with the highest value being applied over all the other ones overwriting them.

You can kind of do this already. The last defined rule "wins". So if you reverse the order of the rules rules (`\\w+:\\s+(?=(\\w+))` first), the whole selection orange. Is this the behaviour you want?

One small issue with that approach is that the "hoverMessage" does display twice over the "TODO: " portion of the text, which is not ideal, but I could see cases where it would be desired. The `ignoreOthers` you mention could fix that issue.

I'm not keen to add that due to the complexity, unless you have some good examples where this solves some real issues.

While on the topic, a similar issue occurs with links, but it's worse than the "hoverMessage" example as two links cannot apply simultaneously. In that case, I think the behaviour should probably be automatic about which one "wins".

**Comment by georgetzlp** [comment]
Thanks for the reply!

> You can kind of do this already. The last defined rule "wins". So if you reverse the order of the rules rules (\\w+:\\s+(?=(\\w+)) first), the whole selection orange. Is this the behaviour you want?

Unfortunately I was after the hoverMessage being displayed twice so that solution wouldn't work.

> I'm not keen to add that due to the complexity, unless you have some good examples where this solves some real issues.

I understand that this is quite complex and I completely understand if you aren't interested in implementing it. It doesn't really have much of a benefit considering the regex could just be modified slightly to have the same effect, it would only just improve the regex you'd have to write slightly.

> While on the topic, a similar issue occurs with links, but it's worse than the "hoverMessage" example as two links cannot apply simultaneously. In that case, I think the behaviour should probably be automatic about which one "wins".

I believe the simplest way to avoid that is to make it so the first regex in the list that matches is the one that get applied considering adding another key you'd need to configure would just create the same effect without any benefit as far as I can tell.

In conclusion I believe something similar to this could improve working with large configs but it's not really necessary considering you could get the same result just by modifying your regex a bit so I completely understand if you don't think this is something worth adding.

**Comment by dlevs** [comment]
I think it's a good idea to add features that reduce the complexity of regexes you need to write.

The order determines the priority, so I think the only thing needed is an `onConflict` property with allowed values `'stack'` (default) or `'replace'`. That's the solution I'm leaning towards.

I don't plan to implement it, but I'd accept a PR for that, or be open to hearing other solutions. :)

**Comment by georgetzlp** [comment]
That sounds like a good solution to me!

I'd love to implement it myself but I don't have any experience in developing VSCode extensions. I might look into this in the future and try implementing it.
